File: leetcode/python/0098. Validate Binary Search Tree.py
# Time complexity: O(N)
# Space complexity: O(H), where H is the height of the tree

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def isValidBST(self, root):
        sequence = []
        self.in_order(sequence, root)
        
        for i in range(1, len(sequence)):
            if sequence[i] <= sequence[i-1]:
                return False
        return True


    def in_order(self, lst, root):
        
        if root != None:
            self.in_order(lst, root.left)
            lst.append(root.val)
            self.in_order(lst, root.right)

File: leetcode/java/0098. Validate Binary Search Tree.java
package leetcode.java;

import java.util.ArrayList;
import java.util.List;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// Time complexity: O(N)
// Space complexity: O(H), where H is the height of the tree

class Solution {
    public boolean isValidBST(TreeNode root) {
        List<Integer> storage = new ArrayList<>();
        this.in_order_traversal(root, storage);
        for (int i=1; i<storage.size(); i++) {
            if (storage.get(i) <= storage.get(i-1)) {
                return false;
            }
        }
        return true;
    }

    public void in_order_traversal(TreeNode root, List<Integer> storage) {
        if (root != null) {
            this.in_order_traversal(root.left, storage);
            storage.add(root.val);
            this.in_order_traversal(root.right, storage);
        }
    }
}